### PR TITLE
Additional tests for async boundaries and GroupBy

### PR DESCRIPTION
**Pull request by johanandren**
A few new tests, some test cleanup, and a tiny docs addition to clarify how `async` works in the context of `groupBy`

Refs #24676

### CODE REVIEW

File: akka-stream-tests/src/test/scala/akka/stream/FusingSpec.scala [hunk 1]
```diff
@@ -69,6 +69,47 @@ class FusingSpec extends StreamSpec {
...
+        .groupBy(10, identity)
+        .via(async)
+        .mergeSubstreams
+        .runWith(Sink.seq)
+        .futureValue
+        .sorted should ===(0 to 9)
+      val refs = receiveN(10)
+      withClue(s"refs=\n${refs.mkString("\n")}") {
+        refs.toSet should have size (11) // main flow + 10 subflows
+      }
```
Comment (johanandren) [inline akka-stream-tests/src/test/scala/akka/stream/FusingSpec.scala]: This becomes just one actor

File: akka-stream-tests/src/test/scala/akka/stream/FusingSpec.scala [hunk 2]
```diff
@@ -69,6 +69,47 @@ class FusingSpec extends StreamSpec {
...
+        .async
+        .mergeSubstreams
+        .runWith(Sink.seq)
+        .futureValue
+        .sorted should ===(0 to 9)
+      val refs = receiveN(10)
+
+      refs.toSet should have size (11) // outer/main flow + 10 subflows
+
+    }
```
Comment (johanandren) [inline akka-stream-tests/src/test/scala/akka/stream/FusingSpec.scala]: this somehow is 10 actors, so one per substream but not a separate one for the stages outside the substream. Very strange.

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowGroupBySpec.scala [hunk 3]
```diff
@@ -590,6 +592,27 @@ class FlowGroupBySpec extends StreamSpec {
...
+      val threeProcessed = TestLatch()
+      val queue = Source.queue[Elem](3, OverflowStrategy.backpressure)
+        .groupBy(2, _.substream)
+        .buffer(2, OverflowStrategy.backpressure)
+        .map { _.f() }.async
+        .to(Sink.ignore)
+        .run()
+
+      List(
+        Elem(1, 1, () ⇒ { Thread.sleep(1.seconds.toMillis); 1 }),
```
Comment (patriknw) [inline akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowGroupBySpec.scala]: might be better to use a CountDownLatch instead

**Comment by johanandren** [comment]
Adding a debug flag I can see that at least in the `subflow.async` case there are actors spawned, so the `FusingSpec` isn't quite telling the truth when it fails.

File: akka-stream-tests/src/test/scala/akka/stream/FusingSpec.scala (hunk 2 above)
Comment (johanandren) [reply akka-stream-tests/src/test/scala/akka/stream/FusingSpec.scala]: Actually today seeing a different behavior here, not sure if I misinterpreted the output before. Now I see one actor for the "main" flow and then each sub stream is added to the same async island.

File: akka-stream/src/main/scala/akka/stream/impl/PhasedFusingActorMaterializer.scala [hunk 4]
```diff
@@ -742,7 +743,11 @@ private final case class SavedIslandData(islandGlobalOffset: Int, lastVisitedOff
           case OptionVal.Some(n) ⇒ n
           case OptionVal.None    ⇒ islandName
         }
-        materializer.actorOf(props, actorName)
+
+        val ref = materializer.actorOf(props, actorName)
+        if (PhasedFusingActorMaterializer.DebugSpawnIsland) {
```
[the remark this answers is not in the document]
Comment (johanandren) [reply akka-stream/src/main/scala/akka/stream/impl/PhasedFusingActorMaterializer.scala]: Let's put it under the same debug flag.

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowGroupBySpec.scala [hunk 5]
```diff
@@ -590,6 +590,29 @@ class FlowGroupBySpec extends StreamSpec {
       upstreamSubscription.sendComplete()
     }
 
+    "not block all substreams when one is blocked but has a buffer in front" in assertAllStagesStopped {
```
[the remark this answers is not in the document]
Comment (johanandren) [reply akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowGroupBySpec.scala]: I don't think so, added some comments to clarify test case

File: akka-stream/src/main/scala/akka/stream/scaladsl/SubFlow.scala [hunk 6]
```diff
@@ -20,6 +20,9 @@ trait SubFlow[+Out, +Mat, +F[+_], C] extends FlowOps[Out, Mat] {
   /**
    * Attach a [[Sink]] to each sub-flow, closing the overall Graph that is being
    * constructed.
+   *
+   * Note that attributes set on the returned graph, including async boundaries are now for the entire graph and not
+   * the `SubFlow`. for example `async` will not have any effect as the returned graph is the entire, closed graph.
```
[the remark this answers is not in the document]
Comment (johanandren) [reply akka-stream/src/main/scala/akka/stream/scaladsl/SubFlow.scala]: Added!

**Review by chbatey — APPROVED** [review]
LGTM

**Review by patriknw — APPROVED** [review]
LGTM

**Comment by patriknw** [comment]
PLS BUILD

**Pull request merged**